tools/get-upload-stats.py

```diff
@@ -0,0 +1,333 @@
+
+
+@dataclass
+class AssetInfo:
+    path: str
+    size: int
+    key: str
+    subject: Optional[str]
+
+
+@dataclass
+class CommitInfo:
+    committish: str
+    short_hash: str
+    created: datetime
+
+
+@dataclass
+class CommitDelta:
+    first: CommitInfo
+    second: CommitInfo
+    assets_added: int
+    assets_added_size: int
+    assets_removed: int
+    assets_removed_size: int
+    duplicates_delta: int
+    duplicates_delta_size: int
+    duplicates_remaining: int
+    subjects_added: int
+
+    def __bool__(self) -> bool:
+        return any(
+            getattr(self, field) != 0
+            for field in [
+                "assets_added",
+                "assets_added_size",
+                "assets_removed",
+                "assets_removed_size",
+                "duplicates_delta",
+                "duplicates_delta_size",
+                # NOT "duplicates_remaining",
+                "subjects_added",
+            ]
+        )
+
+    def for_json(self) -> dict:
+        return {
+            "first": {
+                "committish": self.first.short_hash,
+                "created": self.first.created.isoformat(),
+            },
+            "second": {
+                "committish": self.second.short_hash,
+                "created": self.second.created.isoformat(),
+            },
+            "assets": {
+                "distinct": {
+                    "added": self.assets_added,
+                    "added_size": self.assets_added_size,
+                    "removed": self.assets_removed,
+                    "removed_size": self.assets_removed_size,
+                },
+                "duplicates": {
+                    "delta": self.duplicates_delta,
+                    "delta_size": self.duplicates_delta_size,
+                    "remaining": self.duplicates_remaining,
+                },
+            },
+            "subjects": {
+                "added": self.subjects_added,
+            },
+        }
+
+    def to_markdown(self) -> str:
+        s = f"## Changes from {self.first.short_hash} to {self.second.short_hash}\n\n"
+        s += f"* Assets added: {self.assets_added} ({self.assets_added_size} bytes)\n"
+        s += (
+            f"* Assets removed: {self.assets_removed}"
+            f" ({self.assets_removed_size} bytes)\n"
+        )
+        s += (
+            f"* Duplicates delta: {self.duplicates_delta}"
+            f" ({self.duplicates_delta_size} bytes)\n"
+        )
+        s += f"* Duplicates remaining: {self.duplicates_remaining}\n"
+        s += f"* Subjects added: {self.subjects_added}\n"
+        return s
+
+
+@dataclass
+class Report:
+    from_dt: Optional[datetime]
+    to_dt: Optional[datetime]
+    commit_delta: CommitDelta
+    published_versions: int
+    since_latest: Optional[CommitDelta]
+
+    def for_json(self) -> dict:
+        return {
+            "from_dt": self.from_dt.isoformat() if self.from_dt is not None else None,
```

Comment:
I might have been misspecifying but no example for elderly me is provided, and it seems like it did pick up some date but just incorrect one:

<details>
<summary>my first attempt at it</summary> 

```shell
(git-annex)lena:~/proj/dandi/dandisets[draft]git
$> tools/get-upload-stats.py --from 'Mon Oct 19 23:44:25 2020 -0400' 000003           
# Changes since 2020-10-19 23:44:25-04:00

**Versions published:** 1

No changes since version 0.210812.1448 published on 2021-08-12 14:48:34-04:00

## Changes from 7263a06 to 4ae895f

* Assets added: 0 (0 bytes)
* Assets removed: 0 (0 bytes)
* Duplicates delta: 0 (0 bytes)
* Duplicates remaining: 0
* Subjects added: 0

(dev3) 1 16580 [1].....................................:Tue 21 Sep 2021 02:54:49 PM EDT:.
(git-annex)lena:~/proj/dandi/dandisets[draft]git
$> tools/get-upload-stats.py --from '2020/10/19' 000003           
Traceback (most recent call last):
  File "/home/yoh/proj/dandi/dandisets/tools/get-upload-stats.py", line 333, in <module>
    main()
  File "/home/yoh/proj/dandi/dandisets/venvs/dev3/lib/python3.9/site-packages/click/core.py", line 1137, in __call__
    return self.main(*args, **kwargs)
  File "/home/yoh/proj/dandi/dandisets/venvs/dev3/lib/python3.9/site-packages/click/core.py", line 1062, in main
    rv = self.invoke(ctx)
  File "/home/yoh/proj/dandi/dandisets/venvs/dev3/lib/python3.9/site-packages/click/core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/yoh/proj/dandi/dandisets/venvs/dev3/lib/python3.9/site-packages/click/core.py", line 763, in invoke
    return __callback(*args, **kwargs)
  File "/home/yoh/proj/dandi/dandisets/tools/get-upload-stats.py", line 294, in main
    commit1, commit2 = dd.get_first_and_last_commit(from_dt, to_dt)
  File "/home/yoh/proj/dandi/dandisets/tools/get-upload-stats.py", line 177, in get_first_and_last_commit
    commits = filter_commits(commits, from_dt, to_dt)
  File "/home/yoh/proj/dandi/dandisets/tools/get-upload-stats.py", line 324, in filter_commits
    while commits and commits[0].created < from_dt:
TypeError: can't compare offset-naive and offset-aware datetimes
(dev3) 1 16581 ->1 [1].....................................:Tue 21 Sep 2021 02:55:14 PM EDT:.
(git-annex)lena:~/proj/dandi/dandisets[draft]git
$> git -C 000003 show 7263a06
commit 7263a06eebf7c0b3d2079ab2e603a1aec7dbb441
Author: DANDI Meta-user <dandi@mit.edu>
Date:   Thu May 6 23:59:46 2021 -0400

    [backups2datalad]

diff --git a/dandiset.yaml b/dandiset.yaml
index be7bd63..aaa06ba 100644
--- a/dandiset.yaml
+++ b/dandiset.yaml
@@ -1,7 +1,10 @@
 # DO NOT EDIT THIS FILE LOCALLY. ALL LOCAL UPDATES WILL BE LOST.
 # It can be edited online at https://dandiarchive.org/dandiset/000003
 # and obtained from the dandiarchive.
-about: []
+about:
+- identifier: UBERON:0002436
+  name: primary visual cortex
+  schemaKey: Anatomy
 access:
 - contactPoint:
     email: petersen.peter@gmail.com

```
</details>

<details>
<summary>and even if I set date to 2000 -- still nothing "relevant" was found</summary> 

```shell
$> tools/get-upload-stats.py --from 'Mon Oct 19 23:44:25 2000 -0400' 000003
# Changes since 2000-10-19 23:44:25-04:00

**Versions published:** 1

No changes since version 0.210812.1448 published on 2021-08-12 14:48:34-04:00

## Changes from 7263a06 to 4ae895f

* Assets added: 0 (0 bytes)
* Assets removed: 0 (0 bytes)
* Duplicates delta: 0 (0 bytes)
* Duplicates remaining: 0
* Subjects added: 0
```
</details>

Please 
- add an example of invocation to the output of `--help`
- make sure that if date parsing "didn't work out" - there is an error
- issue a warning (in logger, stderr) if history is "non-linear" (it seems what might have confused on 000003) for the relevant time period
- provide here in PR some exemplar invocations/output on some of our dandisets

Reply:
@yarikoptic

> add an example of invocation to the output of `--help`

What exactly do you need the example to exhibit?  Valid date formats?  The script accepts anything that [`dateutil.parser.parse()`](https://dateutil.readthedocs.io/en/stable/parser.html#dateutil.parser.parse) accepts.

> make sure that if date parsing "didn't work out" - there is an error

There already was an error in your second invocation, because you omitted the timezone information.  Exactly what do you want changed?

Part of the problem with your tests may be that the script (like the version-base search in `backups2datalad.py`) filters out any commits that don't contain "[backups2datalad]", which means that older commits, merge commits, and URL download commits are ignored.  Do you want the script to keep doing that or not?  Note that, if this filter is removed, then there will be a number of commit histories with the author dates out of order, as the author date wasn't always set for URL download commits.

Reply:
> What exactly do you need the example to exhibit? Valid date formats?

yes.

> filters out any commits that don't contain "[backups2datalad]", which means that older commits, merge commits, and URL download commits are ignored. Do you want the script to keep doing that or not? 

I think that makes sense, but may be make it also consider ones which have `Ran backups2datalad.py`?

My point with above was that I got NOTHING reported really on 000003 even though I have specified a date very far in the past.  May be on some elderly dandisets like this indeed we would not be able to get sensible "report".

@satra -- which are BICCN dandisets we should primarily care about ATM, and what would be the reporting date ranges for them?

Reply:
https://gui.dandiarchive.org/#/dandiset/search?search=biccn

Reply:
the next due date is oct 7 i believe. but we can find the submission dates from the past as well.

Reply:
so, next reporting period would be "beginning of the universe" -> "2021/10/07" (unlikely anyone cares about timezone)

Reply:
@yarikoptic I've updated the script to take "Ran backups2datalad.py" commits into consideration, convert naïve datetimes to system time, and provide examples of acceptable timestamp formats.

Reply:
Q2 was 7/8 and Q3 is 10/7, so that delta would be good to compute.

Reply:
Dear @jwodder - could you please produce reports (and see if corresponds to actual changes in git history) for those dandisets for Q2 and Q3 with examples of invocation and actual reports?

Reply:
@yarikoptic Assuming that the intended date ranges are Beginning of Time → July 8, 2021, and Beginning of Time → October 7, 2021, respectively, running the script on 000003 produces:

---

`./get-upload-stats.py --to '2021-07-08 00:00:00 UTC' ../000003`

# Changes up to 2021-07-08 00:00:00+00:00

**Versions published:** 0

## Changes from bbb5209 to 6b56ac0

* Assets added: 100 (2547317560305 bytes)
* Assets removed: 87 (980418516972 bytes)
* Duplicates delta: 0 (0 bytes)
* Duplicates remaining: 0
* Subjects added: 1

---

`./get-upload-stats.py --to '2021-10-07 00:00:00 UTC' ../000003`

# Changes up to 2021-10-07 00:00:00+00:00

**Versions published:** 1

No changes since version 0.210812.1448 published on 2021-08-12 14:48:45+00:00

## Changes from bbb5209 to f2da8ee

* Assets added: 100 (2547317560305 bytes)
* Assets removed: 87 (980418516972 bytes)
* Duplicates delta: 0 (0 bytes)
* Duplicates remaining: 0
* Subjects added: 1

Reply:
For the Q3, start from next day after the end of Q2.

Please use humanize for sizes

Reply:
@yarikoptic Updated outputs:

---

`./get-upload-stats.py --to '2021-07-08 00:00:00 UTC' ../000003`

# Changes up to 2021-07-08 00:00:00+00:00

**Versions published:** 0

## Changes from bbb5209 to 6b56ac0

* Assets added: 100 (2.5 TB)
* Assets removed: 87 (980.4 GB)
* Duplicates delta: 0 (0 Bytes)
* Duplicates remaining: 0
* Subjects added: 1

---

`./get-upload-stats.py --from '2021-07-09 00:00:00 UTC' --to '2021-10-07 00:00:00 UTC' ../000003`

# Changes from 2021-07-09 00:00:00+00:00 to 2021-10-07 00:00:00+00:00

**Versions published:** 1

No changes since version 0.210812.1448 published on 2021-08-12 14:48:45+00:00

## Changes from ab9d37f to f2da8ee

* Assets added: 0 (0 Bytes)
* Assets removed: 0 (0 Bytes)
* Duplicates delta: 0 (0 Bytes)
* Duplicates remaining: 0
* Subjects added: 0